fix: add blocker `-moz` prefix for `:placeholder-shown`

Marukome0743 opened the pull request:
## Description

When Autoprefixer parses the `:placeholder-shown`, the following results are obtained in [Autoprefixer CSS online](https://autoprefixer.github.io/).

![autoprefixer](https://github.com/user-attachments/assets/186d457e-9768-4e4d-aca1-26aaeee549f3)

However, `:-moz-placeholder-shown` doesn't exist at [Mozilla-only properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Mozilla_Extensions#mozilla-only_properties).

![mozilla](https://github.com/user-attachments/assets/75d7ca92-d9a0-45c9-9686-2d8105ee04a8)

That's why I removed the `-moz` prefix from [placeholder-shown.js](https://github.com/postcss/autoprefixer/blob/main/lib/hacks/placeholder-shown.js).

ai commented:
Mozilla had `-moz-` prefix, but called it `:-moz-placeholder`. Can you change `placeholder-shown.js` to use different name on `-moz-`?

https://caniuse.com/css-placeholder-shown

Marukome0743 commented:
> Mozilla had `-moz-` prefix, but called it `:-moz-placeholder`. Can you change `placeholder-shown.js` to use different name on `-moz-`?
> 
> https://caniuse.com/css-placeholder-shown

Thank you for nice review and checking!
I changed the code to use `:-moz-placeholder` instead of `:-moz-placeholder-shown`.
In addition, updated the PR title and description.

ai commented:
Thanks. Released in 10.4.21.

pooledge commented:
Keep in mind that this change effectively breaks it in Firefox (at least for v136@macOS) within the following constellation:
`input:not(:-moz-placeholder) {}`, while 10.4.20 will keep you afloat with `input:not(:-moz-placeholder-shown) {}`

ai commented:
@pooledge can you explain better the problem? Input, output, expected CSS.

pooledge commented:
Input (Svelte scoped SCSS):
```scss
div {
    :global(input) {
        :global(~ label) { /* part 1 */ }
        &:not(:placeholder-shown),
        &:focus {
            :global(~ label) { /* part 2 */ }
        }
    }
}
```

Output v10.4.21 (won't apply for Firefox v136 on mac):
```css
div input:not(:-moz-placeholder) {
  /* part 1 */
}
div input:not(:placeholder-shown), div input:focus {
  /* part 1 */
}
div input:not(:-moz-placeholder) ~ label {
  /* part 2 */
}
div input:not(:placeholder-shown) ~ label, div input:focus ~ label {
  /* part 2 */
}
```

Output v10.4.20 (works well):
```css
div input:not(:-moz-placeholder-shown) {
  /* part 1 */
}
div input:not(:placeholder-shown), div input:focus {
  /* part 1 */
}
div input:not(:-moz-placeholder-shown) ~ label {
  /* part 2 */
}
div input:not(:placeholder-shown) ~ label, div input:focus ~ label {
  /* part 2 */
}
```

ai commented:
> Output v10.4.20 (works well in all browsers):

Why it works well? `div input:not(:-moz-placeholder-shown)` is always `true` on any browser.

ai commented:
@pooledge maybe it was a feature based on bug behavior?

Marukome0743 commented:
I have checked the behavior in FireFox Browser.

<details><summary>HTML code</summary>
<p>

```html
    <main>
      <form class="flex flex-col p-4 w-60">
        <label for="name">Full Name:</label>
        <input id="name" name="name" type="text" />

        <label for="email">Email Address:</label>
        <input
          id="email"
          name="email"
          type="email"
          placeholder="name@example.com"
        />

        <label for="age">Your age:</label>
        <input
          id="age"
          name="age"
          type="number"
          placeholder="You must be 18+"
        />
      </form>
    </main>
```

</p>
</details> 

# ✅`:-placeholder-shown` working

CSS

```css
input:not(:placeholder-shown) {
  background-color: ivory;
  border: 2px solid darkorange;
  border-radius: 5px;
}
```

![placeholder-shown](https://github.com/user-attachments/assets/38225e98-f1e9-4dbf-98c3-b121d375a020)

---

`:-placeholder-shown` only apply input which has placeholder.

# ❌`:-moz-placeholder` working but different behavior

CSS

```css
input:not(:-moz-placeholder) {
  background-color: ivory;
  border: 2px solid darkorange;
  border-radius: 5px;
}
```

![moz-placeholder](https://github.com/user-attachments/assets/ca21ba64-079a-46a5-a194-209c1e16518f)

---

`input:not(:-moz-placeholder)` affect **the all input field** unfortunately.

# ✅`:-moz-placeholder-shown` not working

CSS

```css
input:not(:-moz-placeholder-shown) {
  background-color: ivory;
  border: 2px solid darkorange;
  border-radius: 5px;
}
```

![moz-placeholder-shown](https://github.com/user-attachments/assets/307550b5-0ca2-4556-830b-425efe1fecea)

---

FireFox doesn't recognize `:-moz-placeholder-shown` property because it doesn't exist.

In conclusion, it might be FireFox browser issue.

pooledge commented:
It works because ``

> `div input:not(:-moz-placeholder-shown)` is always `true` on any browser.

..and this is how the non-prefixed styling declaration wins, while with `:-moz-placeholder` it doesn't, since it's a valid selector just like the documentation states. The problem is now a Firefox bug, which makes `:-moz-placeholder` and `:placeholder-shown` work differently, whereas autoprefixer assumes it works exactly the same and prefixes blindly.

I wish `/*! autoprefixer: off */` would work in this case, but it doesn't for me at least in Svelte, though it's a different issue.

pooledge commented:
Update on `/*! autoprefixer: off */`: since it's a `:global()` the only option to apply the comment is to place it in the style root of the component, e.g.:
```svelte
<style lang="scss">
    /*! autoprefixer: off */
    div { /*...*/ }
</style>
```
